# Data Ingestion & Wrangling -- Montgomery County Vision Zero Traffic Progress
Data 205 - Spring 2021  
Mary Glantz

#### Import the libraries needed to import and wrangle the data

In [109]:
import json
import numpy as np
import pandas as pd
import requests



#### Use API's to import data from three Montgomery county datasets (driver info, collision info, and non-motorist collision info) into dataframes.


In [110]:
driverdf = pd.DataFrame(requests.get('https://data.montgomerycountymd.gov/resource/mmzv-x632.json').json())
driverdf.head()

report_number local_case_number                agency_name  \
0    DD5647000F         210007208  Rockville Police Departme   
1    DD5647000F         210007208  Rockville Police Departme   
2    DD5647000C         210007203  Rockville Police Departme   
3   MCP102200FH         210007202   Montgomery County Police   
4   MCP102200FH         210007202   Montgomery County Police   

        acrs_report_type          crash_date_time        route_type  \
0  Property Damage Crash  2021-02-25T10:15:00.000  Maryland (State)   
1  Property Damage Crash  2021-02-25T10:15:00.000  Maryland (State)   
2  Property Damage Crash  2021-02-25T09:16:00.000  Maryland (State)   
3           Injury Crash  2021-02-25T09:16:00.000  Maryland (State)   
4           Injury Crash  2021-02-25T09:16:00.000  Maryland (State)   

          road_name cross_street_type cross_street_name municipality  ...  \
0        NORBECK RD      Municipality         E GUDE DR    ROCKVILLE  ...   
1        NORBECK RD      Municipality         E GUDE DR    ROCKVILLE  ...   
2  W MONTGOMERY AVE      Municipality         NELSON ST    ROCKVILLE  ...   
3          RIDGE RD            County          BRINK RD          N/A  ...   
4          RIDGE RD            County          BRINK RD          N/A  ...   

  :@computed_region_tx5f_5em3 :@computed_region_kbsp_ykn9     latitude  \
0                           3                          25  39.09043867   
1                           3                          25  39.09043867   
2                           3                          25  39.08605617   
3                           2                          39  39.21003667   
4                           2                          39  39.21003667   

      longitude :@computed_region_d7bw_bq6x :@computed_region_rbt8_3x7n  \
0  -77.12867933                          71                          12   
1  -77.12867933                          71                          12   
2  -77.17062833                          71                          12   
3      -77.2399                          36                           1   
4      -77.2399                          36                           1   

                                         geolocation off_road_description  \
0  {'latitude': '39.09043867', 'longitude': '-77....                  NaN   
1  {'latitude': '39.09043867', 'longitude': '-77....                  NaN   
2  {'latitude': '39.08605617', 'longitude': '-77....                  NaN   
3  {'latitude': '39.21003667', 'longitude': '-77....                  NaN   
4  {'latitude': '39.21003667', 'longitude': '-77....                  NaN   

  related_non_motorist non_motorist_substance_abuse  
0                  NaN                          NaN  
1                  NaN                          NaN  
2                  NaN                          NaN  
3                  NaN                          NaN  
4                  NaN                          NaN  

[5 rows x 50 columns]

In [111]:
incidentsdf = pd.DataFrame(requests.get('https://data.montgomerycountymd.gov/resource/bhju-22kf.json').json())
incidentsdf.head()

report_number local_case_number                agency_name  \
0    DD5647000F         210007208  Rockville Police Departme   
1    DD5647000C         210007203  Rockville Police Departme   
2   MCP102200FH         210007202   Montgomery County Police   
3   MCP3036001R         210007162   Montgomery County Police   
4   MCP32180018         210007164   Montgomery County Police   

        acrs_report_type          crash_date_time hit_run        route_type  \
0  Property Damage Crash  2021-02-25T10:15:00.000      No  Maryland (State)   
1  Property Damage Crash  2021-02-25T09:16:00.000      No  Maryland (State)   
2           Injury Crash  2021-02-25T09:16:00.000      No  Maryland (State)   
3  Property Damage Crash  2021-02-24T20:56:00.000      No  Maryland (State)   
4  Property Damage Crash  2021-02-24T20:25:00.000      No            County   

  mile_point mile_point_direction lane_direction  ...  \
0      24.09                 East           West  ...   
1      21.29                 East           West  ...   
2       0.54                North          South  ...   
3       3.76                North          South  ...   
4       2.28                 East           East  ...   

  :@computed_region_tx5f_5em3 :@computed_region_kbsp_ykn9  \
0                           3                          25   
1                           3                          25   
2                           2                          39   
3                           3                          36   
4                           4                          23   

  :@computed_region_d7bw_bq6x :@computed_region_rbt8_3x7n  \
0                          71                          12   
1                          71                          12   
2                          36                           1   
3                          35                          21   
4                          17                           1   

  :@computed_region_a9cs_3ed7 :@computed_region_r648_kzwt        lane_type  \
0                           3                           3              NaN   
1                           3                           3              NaN   
2                           2                           5              NaN   
3                           3                           5  RIGHT TURN LANE   
4                           4                           4              NaN   

  off_road_description related_non_motorist non_motorist_substance_abuse  
0                  NaN                  NaN                          NaN  
1                  NaN                  NaN                          NaN  
2                  NaN                  NaN                          NaN  
3                  NaN                  NaN                          NaN  
4                  NaN                  NaN                          NaN  

[5 rows x 51 columns]

In [112]:
nonmotoristdf = pd.DataFrame(requests.get('https://data.montgomerycountymd.gov/resource/n7fk-dce5.json').json())
nonmotoristdf.head()

report_number local_case_number                agency_name acrs_report_type  \
0    DD56370028         210006695  Rockville Police Departme     Injury Crash   
1   MCP3209002B         210005989   Montgomery County Police     Injury Crash   
2   MCP3274000F         210005911   Montgomery County Police     Injury Crash   
3   MCP229800VL         210005324   Montgomery County Police     Injury Crash   
4   MCP3289000G         210005250   Montgomery County Police     Injury Crash   

           crash_date_time                   off_road_description  \
0  2021-02-21T16:23:00.000  PARKING GARAGE OF 1628 E JEFFERSON ST   
1  2021-02-15T20:20:00.000                                    NaN   
2  2021-02-15T10:49:00.000                                    NaN   
3  2021-02-10T15:16:00.000                                    NaN   
4  2021-02-10T06:35:00.000                                    NaN   

  related_non_motorist           collision_type  weather           light  ...  \
0           PEDESTRIAN                    OTHER    CLEAR             N/A  ...   
1           PEDESTRIAN                    OTHER  RAINING            DUSK  ...   
2           PEDESTRIAN                    OTHER      N/A        DAYLIGHT  ...   
3            BICYCLIST  STRAIGHT MOVEMENT ANGLE   CLOUDY        DAYLIGHT  ...   
4           PEDESTRIAN           SINGLE VEHICLE   CLOUDY  DARK LIGHTS ON  ...   

  :@computed_region_d7bw_bq6x :@computed_region_rbt8_3x7n  \
0                          60                          12   
1                        3065                           1   
2                         111                           1   
3                          39                           1   
4                          21                           1   

  :@computed_region_a9cs_3ed7 :@computed_region_r648_kzwt        route_type  \
0                           3                           3               NaN   
1                           5                           1            County   
2                           1                           3            County   
3                           2                           5            County   
4                           5                           1  Maryland (State)   

            road_name cross_street_type cross_street_name municipality  \
0                 NaN               NaN               NaN          NaN   
1           WAYNE AVE        US (State)       GEORGIA AVE          N/A   
2            BROAD ST            County      GREENLAWN DR          N/A   
3  FATHER HURLEY BLVD            County       WISTERIA DR          N/A   
4   UNIVERSITY BLVD E            County   WILLIAMSBURG DR          N/A   

  surface_condition  
0               NaN  
1               WET  
2               DRY  
3               DRY  
4               DRY  

[5 rows x 39 columns]

In [113]:
marylanddf = pd.DataFrame(requests.get('https://opendata.maryland.gov/resource/65du-s3qu.json').json())
marylanddf.head()

year quarter light_desc light_code     county_desc county_no  \
0  2020      Q2   Daylight          1       Baltimore         3   
1  2020      Q2        NaN       6.02  Baltimore City        24   
2  2020      Q2   Daylight          1      Montgomery        15   
3  2017      Q2   Daylight          1  Baltimore City        24   
4  2020      Q2   Daylight          1           Cecil         7   

       collision_type_desc collision_type_code    fix_obj_desc fix_obj_code  \
0                    Other                  88  Not Applicable            0   
1                    Other                  88      Other Pole           10   
2                    Other                  88  Not Applicable            0   
3           Single Vehicle                  17  Not Applicable            0   
4  Same Direction Rear End                   3  Not Applicable            0   

   ... reference_road_name muni_code rte_no route_type_code  \
0  ...                 NaN       NaN    NaN             NaN   
1  ...           NORTH AVE       NaN    NaN             NaN   
2  ...                 NaN       NaN    NaN             NaN   
3  ...                 NaN       NaN    NaN             NaN   
4  ...                 NaN       NaN    NaN             NaN   

  logmile_dir_flag_desc logmile_dir_flag reference_no reference_type_code  \
0                   NaN              NaN          NaN                 NaN   
1                   NaN              NaN          NaN                 NaN   
2                   NaN              NaN          NaN                 NaN   
3                   NaN              NaN          NaN                 NaN   
4                   NaN              NaN          NaN                 NaN   

  reference_suffix rte_suffix  
0              NaN        NaN  
1              NaN        NaN  
2              NaN        NaN  
3              NaN        NaN  
4              NaN        NaN  

[5 rows x 56 columns]

In [114]:
#crashesDCdf = pd.DataFrame(requests.get('https://opendata.arcgis.com/datasets/70392a096a8e431381f1f692aaa06afd_24.geojson').json())
# crashesDCdf.head()


#### Begin cleaning driver info dataframe  
First, examine the information about the dataframe.


In [115]:
driverdf.shape

(1000, 50)

In [116]:
driverdf.columns

Index(['report_number', 'local_case_number', 'agency_name', 'acrs_report_type',
       'crash_date_time', 'route_type', 'road_name', 'cross_street_type',
       'cross_street_name', 'municipality', 'collision_type', 'weather',
       'surface_condition', 'light', 'traffic_control',
       'driver_substance_abuse', 'person_id', 'driver_at_fault',
       'injury_severity', 'circumstance', 'driver_distracted_by',
       'drivers_license_state', 'vehicle_id', 'vehicle_damage_extent',
       'vehicle_first_impact_location', 'vehicle_second_impact_location',
       'vehicle_body_type', 'vehicle_movement', 'vehicle_continuing_dir',
       'vehicle_going_dir', 'speed_limit', 'driverless_vehicle',
       'parked_vehicle', ':@computed_region_a9cs_3ed7', 'vehicle_year',
       'vehicle_make', ':@computed_region_r648_kzwt', 'vehicle_model',
       ':@computed_region_vu5j_pcmz', 'equipment_problems',
       ':@computed_region_tx5f_5em3', ':@computed_region_kbsp_ykn9',
       'latitude', 'longitude'

In [117]:
driverdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 50 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   report_number                   1000 non-null   object
 1   local_case_number               1000 non-null   object
 2   agency_name                     1000 non-null   object
 3   acrs_report_type                1000 non-null   object
 4   crash_date_time                 1000 non-null   object
 5   route_type                      902 non-null    object
 6   road_name                       906 non-null    object
 7   cross_street_type               902 non-null    object
 8   cross_street_name               906 non-null    object
 9   municipality                    906 non-null    object
 10  collision_type                  1000 non-null   object
 11  weather                         1000 non-null   object
 12  surface_condition               906 non-null    o

The number of nun-null entries indicates whether or not all 1000 entries are non-null.  In several columns, there are fewer than 1000 non-null entries, which means there should be nulls.  The next step is to check for consistent missing value notation.

#### Convert date-time objects to datetime format

In [118]:
driverdf['crash_date_time'] = pd.to_datetime(driverdf['crash_date_time'])
driverdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 50 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   report_number                   1000 non-null   object        
 1   local_case_number               1000 non-null   object        
 2   agency_name                     1000 non-null   object        
 3   acrs_report_type                1000 non-null   object        
 4   crash_date_time                 1000 non-null   datetime64[ns]
 5   route_type                      902 non-null    object        
 6   road_name                       906 non-null    object        
 7   cross_street_type               902 non-null    object        
 8   cross_street_name               906 non-null    object        
 9   municipality                    906 non-null    object        
 10  collision_type                  1000 non-null   object        
 11  weath

#### Check for consistency of variables within columns
I'm not going to check road names because of the expected massive variability within those categories.

In [119]:
reports = driverdf.report_number
#reports_unique = reports.unique()
reports.nunique()

595

This is very interesting, because it indicates that of the 1000 entries, there are only 595 unique reports.  I will check that by using "duplicated" to see how many duplicate entries there are.

In [120]:
duplicateRowsDF = driverdf[driverdf.duplicated('report_number')]
duplicateRowsDF

report_number local_case_number                agency_name  \
1      DD5647000F         210007208  Rockville Police Departme   
4     MCP102200FH         210007202   Montgomery County Police   
6     MCP3036001R         210007162   Montgomery County Police   
8     MCP32180018         210007164   Montgomery County Police   
10     DD5595009H         210007147  Rockville Police Departme   
..            ...               ...                        ...   
988   MCP30390055         210001767   Montgomery County Police   
990   MCP2927003R         210002381   Montgomery County Police   
992   MCP1128005B         210002369   Montgomery County Police   
995   MCP31740031         210002359   Montgomery County Police   
999   MCP2671002J         210002364   Montgomery County Police   

          acrs_report_type     crash_date_time        route_type  \
1    Property Damage Crash 2021-02-25 10:15:00  Maryland (State)   
4             Injury Crash 2021-02-25 09:16:00  Maryland (State)   
6    Property Damage Crash 2021-02-24 20:56:00  Maryland (State)   
8    Property Damage Crash 2021-02-24 20:25:00            County   
10            Injury Crash 2021-02-24 18:25:00  Maryland (State)   
..                     ...                 ...               ...   
988  Property Damage Crash 2021-01-19 16:00:00  Maryland (State)   
990  Property Damage Crash 2021-01-19 15:52:00  Maryland (State)   
992  Property Damage Crash 2021-01-19 15:36:00  Maryland (State)   
995           Injury Crash 2021-01-19 15:28:00            County   
999  Property Damage Crash 2021-01-19 15:18:00            County   

                  road_name cross_street_type    cross_street_name  \
1                NORBECK RD      Municipality            E GUDE DR   
4                  RIDGE RD            County             BRINK RD   
6    MONTGOMERY VILLAGE AVE           Unknown  SPUR TO RUSSELL AVE   
8               RANDOLPH RD  Maryland (State)      CONNECTICUT AVE   
10         W MONTGOMERY AVE      Municipality            NELSON ST   
..                      ...               ...                  ...   
988           EAST WEST HWY  Maryland (State)        COLESVILLE RD   
990       UNIVERSITY BLVD E            County     INDIAN SPRING DR   
992          ROCKVILLE PIKE            County          SECURITY LA   
995         WATKINS MILL RD            County       APPLE RIDGE RD   
999               FENTON ST  Maryland (State)     PHILADELPHIA AVE   

    municipality  ... :@computed_region_tx5f_5em3 :@computed_region_kbsp_ykn9  \
1      ROCKVILLE  ...                           3                          25   
4            N/A  ...                           2                          39   
6            N/A  ...                           3                          36   
8            N/A  ...                           4                          23   
10     ROCKVILLE  ...                           3                          25   
..           ...  ...                         ...                         ...   
988          N/A  ...                           5                           7   
990          N/A  ...                           5                           8   
992          N/A  ...                           1                          19   
995          N/A  ...                           2                          38   
999          N/A  ...                           5                           7   

        latitude     longitude :@computed_region_d7bw_bq6x  \
1    39.09043867  -77.12867933                          71   
4    39.21003667      -77.2399                          36   
6    39.15403167  -77.20922667                          35   
8     39.0572375  -77.07452383                          17   
10   39.08578501  -77.17066705                          71   
..           ...           ...                         ...   
988  38.99352352  -77.03212171                        3065   
990  39.01405612  -77.00419546                          21   
992  39.04210833  -77.

So approximately half of my entries are duplicates.  I will need to figure out what I want to do about that.  I will probably look at the duplicates and see what the differences between the entries are, and then delete or merge the rows as appropriate.



In [121]:
cases = driverdf.local_case_number
#cases_unique = cases.unique()
cases.nunique()

595

In [122]:
duplicateCasesDF = driverdf[driverdf.duplicated('local_case_number')]
duplicateCasesDF

report_number local_case_number                agency_name  \
1      DD5647000F         210007208  Rockville Police Departme   
4     MCP102200FH         210007202   Montgomery County Police   
6     MCP3036001R         210007162   Montgomery County Police   
8     MCP32180018         210007164   Montgomery County Police   
10     DD5595009H         210007147  Rockville Police Departme   
..            ...               ...                        ...   
988   MCP30390055         210001767   Montgomery County Police   
990   MCP2927003R         210002381   Montgomery County Police   
992   MCP1128005B         210002369   Montgomery County Police   
995   MCP31740031         210002359   Montgomery County Police   
999   MCP2671002J         210002364   Montgomery County Police   

          acrs_report_type     crash_date_time        route_type  \
1    Property Damage Crash 2021-02-25 10:15:00  Maryland (State)   
4             Injury Crash 2021-02-25 09:16:00  Maryland (State)   
6    Property Damage Crash 2021-02-24 20:56:00  Maryland (State)   
8    Property Damage Crash 2021-02-24 20:25:00            County   
10            Injury Crash 2021-02-24 18:25:00  Maryland (State)   
..                     ...                 ...               ...   
988  Property Damage Crash 2021-01-19 16:00:00  Maryland (State)   
990  Property Damage Crash 2021-01-19 15:52:00  Maryland (State)   
992  Property Damage Crash 2021-01-19 15:36:00  Maryland (State)   
995           Injury Crash 2021-01-19 15:28:00            County   
999  Property Damage Crash 2021-01-19 15:18:00            County   

                  road_name cross_street_type    cross_street_name  \
1                NORBECK RD      Municipality            E GUDE DR   
4                  RIDGE RD            County             BRINK RD   
6    MONTGOMERY VILLAGE AVE           Unknown  SPUR TO RUSSELL AVE   
8               RANDOLPH RD  Maryland (State)      CONNECTICUT AVE   
10         W MONTGOMERY AVE      Municipality            NELSON ST   
..                      ...               ...                  ...   
988           EAST WEST HWY  Maryland (State)        COLESVILLE RD   
990       UNIVERSITY BLVD E            County     INDIAN SPRING DR   
992          ROCKVILLE PIKE            County          SECURITY LA   
995         WATKINS MILL RD            County       APPLE RIDGE RD   
999               FENTON ST  Maryland (State)     PHILADELPHIA AVE   

    municipality  ... :@computed_region_tx5f_5em3 :@computed_region_kbsp_ykn9  \
1      ROCKVILLE  ...                           3                          25   
4            N/A  ...                           2                          39   
6            N/A  ...                           3                          36   
8            N/A  ...                           4                          23   
10     ROCKVILLE  ...                           3                          25   
..           ...  ...                         ...                         ...   
988          N/A  ...                           5                           7   
990          N/A  ...                           5                           8   
992          N/A  ...                           1                          19   
995          N/A  ...                           2                          38   
999          N/A  ...                           5                           7   

        latitude     longitude :@computed_region_d7bw_bq6x  \
1    39.09043867  -77.12867933                          71   
4    39.21003667      -77.2399                          36   
6    39.15403167  -77.20922667                          35   
8     39.0572375  -77.07452383                          17   
10   39.08578501  -77.17066705                          71   
..           ...           ...                         ...   
988  38.99352352  -77.03212171                        3065   
990  39.01405612  -77.00419546                          21   
992  39.04210833  -77.

I have the same situation with cases, so I assume they are the same---I'll double check by checking both columns.

In [123]:
driverdf[driverdf['local_case_number']=='210002364']

report_number local_case_number               agency_name  \
997   MCP2671002J         210002364  Montgomery County Police   
999   MCP2671002J         210002364  Montgomery County Police   

          acrs_report_type     crash_date_time route_type  road_name  \
997  Property Damage Crash 2021-01-19 15:18:00     County  FENTON ST   
999  Property Damage Crash 2021-01-19 15:18:00     County  FENTON ST   

    cross_street_type cross_street_name municipality  ...  \
997  Maryland (State)  PHILADELPHIA AVE          N/A  ...   
999  Maryland (State)  PHILADELPHIA AVE          N/A  ...   

    :@computed_region_tx5f_5em3 :@computed_region_kbsp_ykn9     latitude  \
997                           5                           7  38.98793019   
999                           5                           7  38.98793019   

       longitude :@computed_region_d7bw_bq6x :@computed_region_rbt8_3x7n  \
997  -77.0235645                        3065                           1   
999  -77.0235645                        3065                           1   

                                           geolocation off_road_description  \
997  {'latitude': '38.98793019', 'longitude': '-77....                  NaN   
999  {'latitude': '38.98793019', 'longitude': '-77....                  NaN   

    related_non_motorist non_motorist_substance_abuse  
997                  NaN                          NaN  
999                  NaN                          NaN  

[2 rows x 50 columns]

In [124]:
driverdf[driverdf['local_case_number']=='210002359']

report_number local_case_number               agency_name  \
994   MCP31740031         210002359  Montgomery County Police   
995   MCP31740031         210002359  Montgomery County Police   

    acrs_report_type     crash_date_time route_type        road_name  \
994     Injury Crash 2021-01-19 15:28:00     County  WATKINS MILL RD   
995     Injury Crash 2021-01-19 15:28:00     County  WATKINS MILL RD   

    cross_street_type cross_street_name municipality  ...  \
994            County    APPLE RIDGE RD          N/A  ...   
995            County    APPLE RIDGE RD          N/A  ...   

    :@computed_region_tx5f_5em3 :@computed_region_kbsp_ykn9     latitude  \
994                           2                          38  39.18419833   
995                           2                          38  39.18419833   

        longitude :@computed_region_d7bw_bq6x :@computed_region_rbt8_3x7n  \
994  -77.21089333                          28                           1   
995  -77.21089333                          28                           1   

                                           geolocation off_road_description  \
994  {'latitude': '39.18419833', 'longitude': '-77....                  NaN   
995  {'latitude': '39.18419833', 'longitude': '-77....                  NaN   

    related_non_motorist non_motorist_substance_abuse  
994                  NaN                          NaN  
995                  NaN                          NaN  

[2 rows x 50 columns]

In these test cases, it appears all of the information matches, and the records are duplicates.  I will remove all duplicates.

In [125]:
driverdf.drop_duplicates(subset=['report_number', 'local_case_number'], keep='last')   # I'm keeping the last rather than the first just in case the last has more recent data.

report_number local_case_number                agency_name  \
1      DD5647000F         210007208  Rockville Police Departme   
2      DD5647000C         210007203  Rockville Police Departme   
4     MCP102200FH         210007202   Montgomery County Police   
6     MCP3036001R         210007162   Montgomery County Police   
8     MCP32180018         210007164   Montgomery County Police   
..            ...               ...                        ...   
993   MCP271700TB         210002365   Montgomery County Police   
995   MCP31740031         210002359   Montgomery County Police   
996   MCP3095002K         210002355   Montgomery County Police   
998   MCP3074004L         210002362   Montgomery County Police   
999   MCP2671002J         210002364   Montgomery County Police   

          acrs_report_type     crash_date_time        route_type  \
1    Property Damage Crash 2021-02-25 10:15:00  Maryland (State)   
2    Property Damage Crash 2021-02-25 09:16:00  Maryland (State)   
4             Injury Crash 2021-02-25 09:16:00  Maryland (State)   
6    Property Damage Crash 2021-02-24 20:56:00  Maryland (State)   
8    Property Damage Crash 2021-02-24 20:25:00            County   
..                     ...                 ...               ...   
993           Injury Crash 2021-01-19 15:31:00  Maryland (State)   
995           Injury Crash 2021-01-19 15:28:00            County   
996  Property Damage Crash 2021-01-19 15:20:00            County   
998  Property Damage Crash 2021-01-19 15:18:00               NaN   
999  Property Damage Crash 2021-01-19 15:18:00            County   

                  road_name cross_street_type    cross_street_name  \
1                NORBECK RD      Municipality            E GUDE DR   
2          W MONTGOMERY AVE      Municipality            NELSON ST   
4                  RIDGE RD            County             BRINK RD   
6    MONTGOMERY VILLAGE AVE           Unknown  SPUR TO RUSSELL AVE   
8               RANDOLPH RD  Maryland (State)      CONNECTICUT AVE   
..                      ...               ...                  ...   
993         CONNECTICUT AVE      Municipality       ROSEMARY ST #2   
995         WATKINS MILL RD            County       APPLE RIDGE RD   
996               ELDRID DR            County          JOHNSON AVE   
998                     NaN               NaN                  NaN   
999               FENTON ST  Maryland (State)     PHILADELPHIA AVE   

       municipality  ... :@computed_region_tx5f_5em3  \
1         ROCKVILLE  ...                           3   
2         ROCKVILLE  ...                           3   
4               N/A  ...                           2   
6               N/A  ...                           3   
8               N/A  ...                           4   
..              ...  ...                         ...   
993  CHEVY CHASE #4  ...                           1   
995             N/A  ...                           2   
996             N/A  ...                           5   
998             NaN  ...                           5   
999             N/A  ...                           5   

    :@computed_region_kbsp_ykn9     latitude     longitude  \
1                            25  39.09043867  -77.12867933   
2                            25  39.08605617  -77.17062833   
4                            39  39.21003667      -77.2399   
6                            36  39.15403167  -77.20922667   
8                            23   39.0572375  -77.07452383   
..                          ...          ...           ...   
993                           5  38.97779433  -77.07710317   
995                          38  39.18419833  -77.21089333   
996                          26     39.06622  -76.99723667   
998                           8    39.000365  -77.00984667   
999                           7  38.98793019   -77.0235645   

    :@computed_region_d7bw_bq6x :@computed_region_rbt8_3x7n  \
1                            71                          12   
2        

Now I will clean up the remaining data.  First checking for entries that need to be merged/changed.  Anytime they use different terms for the same thing (N/A, unknown, etc.), I will replace with one consistent term.

In [126]:
agency_name = driverdf.agency_name
agency_uniques = agency_name.unique()
agency_uniques

array(['Rockville Police Departme', 'Montgomery County Police',
       'Maryland-National Capital', 'Gaithersburg Police Depar',
       'Takoma Park Police Depart'], dtype=object)

This has all consistent terms, but I need to correct the typos caused by shortening the names.

In [127]:
# Using replace by dictionary to correct the names

driverdf['agency_name'] = driverdf['agency_name'].replace({'Rockville Police Departme':'Rockville Police', 'Maryland-National Capital':'Maryland-National Capital Police', 
                                                     'Gaithersburg Police Depar':'Gaithersburg Police', 'Takoma Park Police Depart': 'Takoma Park Police'})
agency_uniques = agency_name.unique()
agency_uniques

array(['Rockville Police', 'Montgomery County Police',
       'Maryland-National Capital Police', 'Gaithersburg Police',
       'Takoma Park Police'], dtype=object)

In [128]:
acrs_report_type = driverdf.acrs_report_type
acrs_uniques = acrs_report_type.unique()
acrs_uniques

array(['Property Damage Crash', 'Injury Crash'], dtype=object)

In [129]:
route_type = driverdf.route_type
route_uniques = route_type.unique()
route_uniques

array(['Maryland (State)', 'County', nan, 'Municipality', 'US (State)',
       'Interstate (State)', 'Other Public Roadway', 'Ramp', 'Government'],
      dtype=object)

There is only one type of unknown here (nan), so I will leave it.

In [130]:
cross_street_type = driverdf.cross_street_type
cross_street_uniques = cross_street_type.unique()
cross_street_uniques

array(['Municipality', 'County', 'Unknown', 'Maryland (State)', nan,
       'US (State)', 'Ramp', 'Other Public Roadway', 'Service Road',
       'Interstate (State)'], dtype=object)

'Unknown' and 'nan' should be combined.

In [131]:
driverdf['cross_street_type'] = driverdf['cross_street_type'].fillna('Unknown')
cross_street_type = driverdf.cross_street_type
cross_street_uniques = cross_street_type.unique()
cross_street_uniques

array(['Municipality', 'County', 'Unknown', 'Maryland (State)',
       'US (State)', 'Ramp', 'Other Public Roadway', 'Service Road',
       'Interstate (State)'], dtype=object)

In [132]:
municipality = driverdf.municipality
municipality_uniques = municipality.unique()
municipality_uniques

array(['ROCKVILLE', 'N/A', nan, 'GAITHERSBURG', 'SOMERSET', 'KENSINGTON',
       'TAKOMA PARK', 'POOLESVILLE', 'FRIENDSHIP HEIGHTS',
       'CHEVY CHASE #4'], dtype=object)

'N/A' and 'nan' should be combined.

In [133]:
driverdf['municipality'] = driverdf['municipality'].fillna('N/A')
municipality = driverdf.municipality
municipality_uniques = municipality.unique()
municipality_uniques

array(['ROCKVILLE', 'N/A', 'GAITHERSBURG', 'SOMERSET', 'KENSINGTON',
       'TAKOMA PARK', 'POOLESVILLE', 'FRIENDSHIP HEIGHTS',
       'CHEVY CHASE #4'], dtype=object)

In [134]:
collision_type = driverdf.collision_type
collision_type_uniques = collision_type.unique()
collision_type_uniques

array(['SAME DIRECTION SIDESWIPE', 'SINGLE VEHICLE',
       'STRAIGHT MOVEMENT ANGLE', 'SAME DIR REAR END',
       'HEAD ON LEFT TURN', 'OTHER', 'HEAD ON',
       'SAME DIRECTION RIGHT TURN', 'OPPOSITE DIRECTION SIDESWIPE',
       'UNKNOWN', 'ANGLE MEETS LEFT HEAD ON', 'ANGLE MEETS RIGHT TURN',
       'OPPOSITE DIR BOTH LEFT TURN', 'ANGLE MEETS LEFT TURN',
       'SAME DIR BOTH LEFT TURN', 'SAME DIRECTION LEFT TURN',
       'SAME DIR REND RIGHT TURN', 'SAME DIR REND LEFT TURN', 'N/A'],
      dtype=object)

In [135]:
# Replace 'N/A' with 'UNKNOWN'
driverdf['collision_type'] = driverdf['collision_type'].replace({'N/A': 'UNKNOWN'})
collision_type = driverdf.collision_type
collision_type_uniques = collision_type.unique()
collision_type_uniques


array(['SAME DIRECTION SIDESWIPE', 'SINGLE VEHICLE',
       'STRAIGHT MOVEMENT ANGLE', 'SAME DIR REAR END',
       'HEAD ON LEFT TURN', 'OTHER', 'HEAD ON',
       'SAME DIRECTION RIGHT TURN', 'OPPOSITE DIRECTION SIDESWIPE',
       'UNKNOWN', 'ANGLE MEETS LEFT HEAD ON', 'ANGLE MEETS RIGHT TURN',
       'OPPOSITE DIR BOTH LEFT TURN', 'ANGLE MEETS LEFT TURN',
       'SAME DIR BOTH LEFT TURN', 'SAME DIRECTION LEFT TURN',
       'SAME DIR REND RIGHT TURN', 'SAME DIR REND LEFT TURN'],
      dtype=object)

In [136]:
weather = driverdf.weather
weather_uniques = weather.unique()
weather_uniques

array(['CLEAR', 'N/A', 'CLOUDY', 'RAINING', 'SLEET', 'SNOW', 'WINTRY MIX',
       'UNKNOWN', 'OTHER', 'BLOWING SNOW', 'FOGGY'], dtype=object)

In [137]:
# Replace 'N/A' with 'UNKNOWN'  

driverdf['weather'] = driverdf['weather'].replace({'N/A':'UNKNOWN'})
weather = driverdf.weather
weather_uniques = weather.unique()
weather_uniques


array(['CLEAR', 'UNKNOWN', 'CLOUDY', 'RAINING', 'SLEET', 'SNOW',
       'WINTRY MIX', 'OTHER', 'BLOWING SNOW', 'FOGGY'], dtype=object)

In [138]:
surface_condition = driverdf.surface_condition
surface_condition_uniques = surface_condition.unique()
surface_condition_uniques

array(['DRY', 'WET', nan, 'SNOW', 'SLUSH', 'UNKNOWN', 'ICE', 'N/A',
       'WATER(STANDING/MOVING)'], dtype=object)

I will combine "nan", "N/A", and "UNKNOWN", and replace nan with "UNKNOWN".

In [139]:
driverdf['surface_condition'] = driverdf['surface_condition'].replace({'N/A': 'UNKNOWN'})
driverdf['surface_condition'] = driverdf['surface_condition'].fillna('UNKNOWN')
surface_condition = driverdf.surface_condition
surface_condition_uniques = surface_condition.unique()
surface_condition_uniques

array(['DRY', 'WET', 'UNKNOWN', 'SNOW', 'SLUSH', 'ICE',
       'WATER(STANDING/MOVING)'], dtype=object)

In [140]:
light = driverdf.light
light_uniques = light.unique()
light_uniques

array(['DAYLIGHT', 'DARK LIGHTS ON', 'DUSK', 'DARK NO LIGHTS', 'UNKNOWN',
       'DAWN', 'N/A', 'DARK -- UNKNOWN LIGHTING', 'OTHER'], dtype=object)

In [141]:
# Replace 'N/A' with 'UNKNOWN'
driverdf['light'] = driverdf['light'].replace({'N/A':'UNKNOWN'})
light = driverdf.light
light_uniques = light.unique()
light_uniques

array(['DAYLIGHT', 'DARK LIGHTS ON', 'DUSK', 'DARK NO LIGHTS', 'UNKNOWN',
       'DAWN', 'DARK -- UNKNOWN LIGHTING', 'OTHER'], dtype=object)

In [142]:
traffic_control = driverdf.traffic_control
traf_ctrl_uniques = traffic_control.unique()
traf_ctrl_uniques

array(['N/A', 'TRAFFIC SIGNAL', 'NO CONTROLS', 'STOP SIGN', 'YIELD SIGN',
       'FLASHING TRAFFIC SIGNAL', 'OTHER', 'UNKNOWN'], dtype=object)

In [143]:
# Replace 'N/A' with 'UNKNOWN'
driverdf['traffic_control'] = driverdf['traffic_control'].replace({'N/A': 'UNKNOWN'})
traffic_control = driverdf.traffic_control
traf_ctrl_uniques = traffic_control.unique()
traf_ctrl_uniques

array(['UNKNOWN', 'TRAFFIC SIGNAL', 'NO CONTROLS', 'STOP SIGN',
       'YIELD SIGN', 'FLASHING TRAFFIC SIGNAL', 'OTHER'], dtype=object)

In [144]:
substance_abuse = driverdf.driver_substance_abuse
substance_uniques = substance_abuse.unique()
substance_uniques


array(['N/A', 'NONE DETECTED', 'ALCOHOL PRESENT', 'UNKNOWN',
       'ALCOHOL CONTRIBUTED', 'OTHER', 'ILLEGAL DRUG PRESENT',
       'MEDICATION CONTRIBUTED', 'MEDICATION PRESENT',
       'COMBINATION CONTRIBUTED'], dtype=object)

In [145]:
# Replace "N/A" with "UNKNOWN"
driverdf['driver_substance_abuse'] = driverdf['driver_substance_abuse'].replace({'N/A': 'UNKNOWN'})
substance_abuse = driverdf.driver_substance_abuse
substance_uniques = substance_abuse.unique()
substance_uniques

array(['UNKNOWN', 'NONE DETECTED', 'ALCOHOL PRESENT',
       'ALCOHOL CONTRIBUTED', 'OTHER', 'ILLEGAL DRUG PRESENT',
       'MEDICATION CONTRIBUTED', 'MEDICATION PRESENT',
       'COMBINATION CONTRIBUTED'], dtype=object)

In [146]:
driver_at_fault = driverdf.driver_at_fault
at_fault_uniques = driver_at_fault.unique()
at_fault_uniques

array(['Yes', 'No', 'Unknown'], dtype=object)

In [147]:
injury_severity = driverdf.injury_severity 
injury_severity_uniques = injury_severity.unique()
injury_severity_uniques

array(['NO APPARENT INJURY', 'SUSPECTED MINOR INJURY', 'POSSIBLE INJURY',
       'SUSPECTED SERIOUS INJURY'], dtype=object)

In [148]:
circumstance = driverdf.circumstance
circumstance_uniques = circumstance.unique()
circumstance_uniques

array(['N/A',
       'PHYSICAL OBSTRUCTION(S), VISION OBSTRUCTION (INCL. BLINDED BY SUN)',
       'ANIMAL, N/A', 'ICY OR SNOW-COVERED, N/A, WET', 'N/A, WET',
       'ICY OR SNOW-COVERED, RAIN, SNOW', 'RAIN, SNOW, WET',
       'N/A, SLEET, HAIL, FREEZ. RAIN',
       'ICY OR SNOW-COVERED, SLEET, HAIL, FREEZ. RAIN',
       'SLEET, HAIL, FREEZ. RAIN, WET',
       'ICY OR SNOW-COVERED, RAIN, SNOW, SLEET, HAIL, FREEZ. RAIN',
       'N/A, VISION OBSTRUCTION (INCL. BLINDED BY SUN)',
       'ICY OR SNOW-COVERED, N/A',
       'V EXHAUST SYSTEM|R OTHER ROAD, V WIPERS|W OTHER ENVIRONMENTAL',
       'ICY OR SNOW-COVERED, SLEET, HAIL, FREEZ. RAIN, WET',
       'ICY OR SNOW-COVERED, RAIN, SNOW, SLEET, HAIL, FREEZ. RAIN, WET',
       'DEBRIS OR OBSTRUCTION, N/A',
       'ICY OR SNOW-COVERED, V WIPERS|W OTHER ENVIRONMENTAL',
       'ICY OR SNOW-COVERED, RAIN, SNOW, WET',
       'N/A, V EXHAUST SYSTEM|R OTHER ROAD', 'N/A, RAIN, SNOW',
       'RAIN, SNOW, V EXHAUST SYSTEM|R OTHER ROAD'], dtype=object)

In [149]:
driver_distracted_by = driverdf.driver_distracted_by
driver_distracted_by_uniques = driver_distracted_by.unique()
driver_distracted_by_uniques

array(['LOOKED BUT DID NOT SEE', 'NOT DISTRACTED',
       'INATTENTIVE OR LOST IN THOUGHT', 'UNKNOWN', 'OTHER DISTRACTION',
       'DISTRACTED BY OUTSIDE PERSON OBJECT OR EVENT',
       'BY MOVING OBJECT IN VEHICLE', 'NO DRIVER PRESENT',
       'EATING OR DRINKING', 'USING DEVICE OBJECT BROUGHT INTO VEHICLE',
       'OTHER CELLULAR PHONE RELATED',
       'OTHER ELECTRONIC DEVICE (NAVIGATIONAL PALM PILOT)',
       'TEXTING FROM A CELLULAR PHONE', 'DIALING CELLULAR PHONE',
       'BY OTHER OCCUPANTS'], dtype=object)

In [150]:
drivers_license_state = driverdf.drivers_license_state
drivers_license_state_uniques = drivers_license_state.unique()
drivers_license_state_uniques

array(['MD', 'CA', nan, 'DC', 'DE', 'NC', 'VA', 'XX', 'PA', 'CT', 'NJ',
       'NY', 'GA', 'WV', 'CO', 'AZ', 'MH', 'MA', 'FL', 'MI', 'SC', 'OH',
       'ME', 'RI', 'TN', 'LA', 'IL'], dtype=object)

In [151]:
# 'XX' appears to be the same as nan---I'll replace nan with 'XX'
driverdf['drivers_license_state'] = driverdf['drivers_license_state'].fillna('XX')
drivers_license_state = driverdf.drivers_license_state
drivers_license_state_uniques = drivers_license_state.unique()
drivers_license_state_uniques

array(['MD', 'CA', 'XX', 'DC', 'DE', 'NC', 'VA', 'PA', 'CT', 'NJ', 'NY',
       'GA', 'WV', 'CO', 'AZ', 'MH', 'MA', 'FL', 'MI', 'SC', 'OH', 'ME',
       'RI', 'TN', 'LA', 'IL'], dtype=object)

In [152]:
vehicle_damage_extent = driverdf.vehicle_damage_extent
vehicle_damage_extent_uniques = vehicle_damage_extent.unique()
vehicle_damage_extent_uniques

array(['DISABLING', 'FUNCTIONAL', 'SUPERFICIAL', 'DESTROYED', 'UNKNOWN',
       'NO DAMAGE', 'OTHER', 'N/A'], dtype=object)

In [153]:
driverdf['vehicle_damage_extent'] = driverdf['vehicle_damage_extent'].replace({'N/A':'UNKNOWN'})
vehicle_damage_extent = driverdf.vehicle_damage_extent
vehicle_damage_extent_uniques = vehicle_damage_extent.unique()
vehicle_damage_extent_uniques

array(['DISABLING', 'FUNCTIONAL', 'SUPERFICIAL', 'DESTROYED', 'UNKNOWN',
       'NO DAMAGE', 'OTHER'], dtype=object)

In [154]:
vehicle_first_impact_location = driverdf.vehicle_first_impact_location
vehicle_first_impact_location_uniques = vehicle_first_impact_location.unique()
vehicle_first_impact_location_uniques

array(['ONE OCLOCK', 'SEVEN OCLOCK', 'TWO OCLOCK', 'TWELVE OCLOCK',
       'ELEVEN OCLOCK', 'SIX OCLOCK', 'FIVE OCLOCK', 'EIGHT OCLOCK',
       'TEN OCLOCK', 'FOUR OCLOCK', 'THREE OCLOCK', 'ROOF TOP',
       'UNDERSIDE', 'NINE OCLOCK', 'UNKNOWN', 'NON-COLLISION'],
      dtype=object)

In [155]:
vehicle_second_impact_location = driverdf.vehicle_second_impact_location
vehicle_second_impact_location_uniques = vehicle_second_impact_location.unique()
vehicle_second_impact_location_uniques

array(['ONE OCLOCK', 'SEVEN OCLOCK', 'TWO OCLOCK', 'TWELVE OCLOCK',
       'ELEVEN OCLOCK', 'SIX OCLOCK', 'FIVE OCLOCK', 'EIGHT OCLOCK',
       'TEN OCLOCK', 'FOUR OCLOCK', 'THREE OCLOCK', 'ROOF TOP',
       'UNDERSIDE', 'NINE OCLOCK', 'UNKNOWN', 'NON-COLLISION'],
      dtype=object)

In [156]:
vehicle_body_type = driverdf.vehicle_body_type
vehicle_body_type_uniques = vehicle_body_type.unique()
vehicle_body_type_uniques

array(['PASSENGER CAR', 'VAN', 'PICKUP TRUCK', 'TRANSIT BUS',
       '(SPORT) UTILITY VEHICLE', 'FIRE VEHICLE/NON EMERGENCY',
       'SCHOOL BUS',
       'CARGO VAN/LIGHT TRUCK 2 AXLES (OVER 10,000LBS (4,536 KG))',
       'POLICE VEHICLE/EMERGENCY',
       'MEDIUM/HEAVY TRUCKS 3 AXLES (OVER 10,000LBS (4,536KG))',
       'POLICE VEHICLE/NON EMERGENCY', 'STATION WAGON',
       'OTHER LIGHT TRUCKS (10,000LBS (4,536KG) OR LESS)',
       'AMBULANCE/EMERGENCY', 'TRUCK TRACTOR', 'UNKNOWN',
       'FIRE VEHICLE/EMERGENCY', 'OTHER', nan, 'MOTORCYCLE', 'OTHER BUS',
       'RECREATIONAL VEHICLE', 'MOPED', 'AMBULANCE/NON EMERGENCY'],
      dtype=object)

In [157]:
# Replace nan with 'UNKNOWN'
driverdf['vehicle_body_type'] = driverdf['vehicle_body_type'].fillna('UNKNOWN')
vehicle_body_type = driverdf.vehicle_body_type
vehicle_body_type_uniques = vehicle_body_type.unique()
vehicle_body_type_uniques

array(['PASSENGER CAR', 'VAN', 'PICKUP TRUCK', 'TRANSIT BUS',
       '(SPORT) UTILITY VEHICLE', 'FIRE VEHICLE/NON EMERGENCY',
       'SCHOOL BUS',
       'CARGO VAN/LIGHT TRUCK 2 AXLES (OVER 10,000LBS (4,536 KG))',
       'POLICE VEHICLE/EMERGENCY',
       'MEDIUM/HEAVY TRUCKS 3 AXLES (OVER 10,000LBS (4,536KG))',
       'POLICE VEHICLE/NON EMERGENCY', 'STATION WAGON',
       'OTHER LIGHT TRUCKS (10,000LBS (4,536KG) OR LESS)',
       'AMBULANCE/EMERGENCY', 'TRUCK TRACTOR', 'UNKNOWN',
       'FIRE VEHICLE/EMERGENCY', 'OTHER', 'MOTORCYCLE', 'OTHER BUS',
       'RECREATIONAL VEHICLE', 'MOPED', 'AMBULANCE/NON EMERGENCY'],
      dtype=object)

In [158]:
vehicle_movement = driverdf.vehicle_movement
vehicle_movement_uniques = vehicle_movement.unique()
vehicle_movement_uniques

array(['MOVING CONSTANT SPEED', 'MAKING LEFT TURN', 'MAKING RIGHT TURN',
       'STARTING FROM PARKED', 'SLOWING OR STOPPING',
       'STOPPED IN TRAFFIC LANE', 'CHANGING LANES', 'ACCELERATING',
       'PARKING', 'BACKING', 'STARTING FROM LANE', 'PARKED', 'UNKNOWN',
       'SKIDDING', 'PASSING', 'OTHER', 'N/A', 'NEGOTIATING A CURVE',
       'MAKING U TURN', 'LEAVING TRAFFIC LANE', 'RIGHT TURN ON RED',
       'ENTERING TRAFFIC LANE'], dtype=object)

In [159]:
vehicle_continuing_dir = driverdf.vehicle_continuing_dir
vehicle_continuing_dir_uniques = vehicle_continuing_dir.unique()
vehicle_continuing_dir_uniques

array(['West', 'South', 'North', 'East', nan, 'Unknown'], dtype=object)

In [160]:
# Replace nan with 'UNKNOWN'
driverdf['vehicle_continuing_dir'] = driverdf['vehicle_continuing_dir'].fillna('UNKNOWN')
driverdf['vehicle_continuing_dir'] = driverdf['vehicle_continuing_dir'].replace({'Unknown':'UNKNOWN'})
vehicle_continuing_dir = driverdf.vehicle_continuing_dir
vehicle_continuing_dir_uniques = vehicle_continuing_dir.unique()
vehicle_continuing_dir_uniques

array(['West', 'South', 'North', 'East', 'UNKNOWN'], dtype=object)

In [161]:
vehicle_going_dir = driverdf.vehicle_going_dir
vehicle_going_dir_uniques = vehicle_going_dir.unique()
vehicle_going_dir_uniques

array(['West', 'South', 'East', 'North', nan, 'Unknown'], dtype=object)

In [162]:
# Replace nan with 'UNKNOWN'
driverdf['vehicle_going_dir'] = driverdf['vehicle_going_dir'].fillna('UNKNOWN')
driverdf['vehicle_going_dir'] = driverdf['vehicle_going_dir'].replace({'Unknown':'UNKNOWN'})
vehicle_going_dir = driverdf.vehicle_going_dir
vehicle_going_dir_uniques = vehicle_going_dir.unique()
vehicle_going_dir_uniques

array(['West', 'South', 'East', 'North', 'UNKNOWN'], dtype=object)

In [163]:
speed_limit = driverdf.speed_limit
speed_limit_uniques = speed_limit.unique()
speed_limit_uniques

array(['50', '30', '40', '35', '15', '25', '5', '45', '10', '0', '20',
       '55', '60'], dtype=object)

In [164]:
parked_vehicle = driverdf.parked_vehicle
parked_vehicle_uniques = parked_vehicle.unique()
parked_vehicle_uniques

array(['No', 'Yes'], dtype=object)

In [165]:
vehicle_year = driverdf.vehicle_year
vehicle_year_uniques = vehicle_year.unique()
vehicle_year_uniques

array(['2006', '2014', '2015', '2016', '2017', '2007', '2009', '2018',
       '2013', '2010', '1996', '1999', '2005', '2020', '2019', '2008',
       '2021', '2012', '2011', '2003', '1997', '2001', '1998', '2002',
       '2004', '2000', '1989', '0', '1986', '1901', '1992', '1995',
       '1990', '2022'], dtype=object)

In [166]:
vehicle_make = driverdf.vehicle_make
vehicle_make_uniques = vehicle_make.unique()
vehicle_make_uniques

array(['TOYOTA', 'TOYT', 'BMW', 'FORD', 'NISSAN', 'HONDA', 'NEW FLYER',
       'MERZ', 'MAZDA', 'INFI', 'CHEV', 'HYUN', 'SUBARU', 'HYUND',
       'THOMAS', 'CHEVROLET', 'JEEP', 'TRANSIT', 'LEXUS', 'CHEVY', 'MACK',
       'HOND', 'LEXU', 'VOLK', 'NISS', 'DODGE', 'FREIGHTLINER', 'KIA',
       'RANGE', 'HYUNDAI', 'DODG', 'NISAN', 'GMC', 'HYUD', 'MERCEDES',
       'ACURA', 'MERC', 'VOLVO', 'LINCOLN', 'INTERNATIONAL', 'AUDI',
       'MNNI', 'UNKNOWN', 'PRTA', 'KENWTH', 'KAWASAKI', 'MAZD', 'PTRB',
       'FRHT', 'RAM', 'NOVABUS', 'TBU', 'GILL', 'KENWORTH', 'MITSUBISHI',
       'MITSUBUSHI', 'BUICK', 'MASERATTI', 'HUYN', 'PIERCE', 'NEWFLYER',
       'VOLKWAGEN', 'INFINITI', 'SUBA', 'LANDOVER', 'TOYO', 'PETE',
       'ISUZU', 'MITS', 'CADI', 'PORSCHE', 'NWFL', 'INTL', 'CHEVORLET',
       'VW', 'SPARTAN', 'CHRY', 'MERCEDES BENZ', 'VOLV', 'TESLA',
       'CHRYSLER', 'SATURN', 'LANDROVER', 'LEXIS', 'VOLKWAGON',
       'VOLKSWAGEN', 'PIER', 'LAND ROVER', 'FRAIGHT LINER', 'NEW', 'SPAR',
       'HON

In [167]:
# Standardize make names --- unfortunately reviewing the make and model together reveals merc can be either mercedes or mercury...
# Some of these abbreviations are mysteries.
driverdf['vehicle_make'] = driverdf['vehicle_make'].replace({'TOYT': 'TOYOTA', 'MERZ': 'MERCEDES', 'INFI': 'INFINITI', 'HYUN': 'HYUNDAI', 'CHEV': 'CHEVROLET',
                                                             'HYUND': 'HYUNDAI', 'CHEVY': 'CHEVROLET', 'LEXU': 'LEXUS', 'HOND': 'HONDA', 'VOLK': 'VOLKSWAGEN',
                                                             'NISS': 'NISSAN', 'RANGE': 'RANGE_ROVER', 'DODG': 'DODGE', 'NISAN': 'NISSAN', 'HYUD': 'HYUNDAI',
                                                             'MNNI': 'MINI', 'MAZD': 'MAZDA', 'MITSUBUSHI': 'MITSUBISHI', 'HUYN': 'HYUNDAI', 'VOLKWAGEN': 'VOLKSWAGEN', 
                                                             'SUBA': 'SUBARU', 'LANDOVER': 'LANDROVER', 'TOYO': 'TOYOTA', 'MITS': 'MITSUBISHI', 'CADI': 'CADILLAC',
                                                             'CHEVORLET': 'CHEVROLET', 'VW': 'VOLKSWAGEN', 'CHRY':'CHRYSLER', 'VOLKWAGON': 'VOLKSWAGEN', 'LAND ROVER': 'LANDROVER', 
                                                             'FRAIGHT LINER': 'FREIGHTLINER', 'HONDM': 'HONDA', 'TOYTA': 'TOYOTA', 'SSUBARU': 'SUBARU', 'VOLKSWAGON': 'VOLKSWAGEN', 'XX': 'UNKNOWN',
                                                             'ACUR': 'ACURA', 'BUICCK': 'BUICK', 'HYNDAI': 'HYUNDAI', 'UNK': 'UNKNOWN', 'HYUNDA':'HYUNDAI', 'SUB': 'SUBARU',
                                                             'Unknown':'UNKNOWN'})
vehicle_make = driverdf.vehicle_make
vehicle_make_uniques = vehicle_make.unique()
vehicle_make_uniques

array(['TOYOTA', 'BMW', 'FORD', 'NISSAN', 'HONDA', 'NEW FLYER',
       'MERCEDES', 'MAZDA', 'INFINITI', 'CHEVROLET', 'HYUNDAI', 'SUBARU',
       'THOMAS', 'JEEP', 'TRANSIT', 'LEXUS', 'MACK', 'VOLKSWAGEN',
       'DODGE', 'FREIGHTLINER', 'KIA', 'RANGE_ROVER', 'GMC', 'ACURA',
       'MERC', 'VOLVO', 'LINCOLN', 'INTERNATIONAL', 'AUDI', 'MINI',
       'UNKNOWN', 'PRTA', 'KENWTH', 'KAWASAKI', 'PTRB', 'FRHT', 'RAM',
       'NOVABUS', 'TBU', 'GILL', 'KENWORTH', 'MITSUBISHI', 'BUICK',
       'MASERATTI', 'PIERCE', 'NEWFLYER', 'LANDROVER', 'PETE', 'ISUZU',
       'CADILLAC', 'PORSCHE', 'NWFL', 'INTL', 'SPARTAN', 'CHRYSLER',
       'MERCEDES BENZ', 'VOLV', 'TESLA', 'SATURN', 'LEXIS', 'PIER', 'NEW',
       'SPAR', 'GILG', 'LLOYDS', 'HINO', 'POSTAL TRUCK', 'STLG', 'VESPA',
       'PONTIAC', 'DIESEL URBAN TRANSIT'], dtype=object)

In [168]:
vehicle_model = driverdf.vehicle_model
vehicle_model_uniques = vehicle_model.unique()
vehicle_model_uniques

array(['4S', 'M4S', 'SIENNA', 'F-150', 'ALTIMA', '328I', 'CIVIC', 'TBU',
       'TK', 'COLORADO', 'ACCORD CROSSTOUR', 'CROSSTREK', 'ELANTRA',
       'CAMRY', 'SCHOOL BUS', 'VN', 'CRV', 'TUNDRA', 'SENTRA', 'TACOMA',
       'BUS', 'COROLLA', 'SEQUOIA', 'IS 300', 'SILVERADO', 'CAMARO',
       'F150', 'CN', '500', 'HR-V', 'ACCORD', 'JETTA', 'SUV', 'ODYSSEY',
       'JOURNEY', 'FOCUS', 'COBALT', 'EXPRESS', 'FORESTER', 'MALIBU',
       'CX9', 'TRUCK', 'SPECTRA', 'OPTIMA', 'EVOQUE', '4D', 'FORTE',
       'RAV4', 'JX35', 'CHARGER', 'PATRIOT TK', 'ALTIMA 4S', 'PATHFINDER',
       'CANYON', 'TUCSON', 'EXPLORER', 'LEGACY', 'C300L', 'TL', '4 DOOR',
       'C5500', 'ELEMENT', 'SW', 'ROGUE', 'TUNDA', '4RUNNER', 'MURANO',
       'VOLT', 'ESCAPE', 'WRANGLER', 'GLE350', 'SD', 'HRV', 'PILOT',
       'DUMP TRUCK', 'XC30', 'ML 350', 'EQUINOX LT', 'PRIUS', 'SQ5',
       'SUBURBAN', 'NAVIGATOR', '1500', 'COOPER', 'MUSTANG', 'M2 106',
       'PICK UP', 'UNKNOWN', 'BA', 'TITAN', 'WT 1500', '4300 R/B',
       

In [169]:
equipment_problems = driverdf.equipment_problems
equipment_problems_uniques = equipment_problems.unique()
equipment_problems_uniques

array(['N/A', 'NO MISUSE', 'FACING WRONG WAY', 'UNKNOWN', 'OTHER'],
      dtype=object)

NOTE:  This column doesn't seem to include any information of interest, so I'll delete it.

In [170]:
related_non_motorist = driverdf.related_non_motorist
related_non_motorist_uniques = related_non_motorist.unique()
related_non_motorist_uniques

array([nan, 'PEDESTRIAN', 'BICYCLIST', 'MACHINE OPERATOR/RIDER'],
      dtype=object)

In [171]:
# Fill na's with "UNKNOWN"
driverdf['related_non_motorist'] = driverdf['related_non_motorist'].fillna("UNKNOWN")
related_non_motorist = driverdf.related_non_motorist
related_non_motorist_uniques = related_non_motorist.unique()
related_non_motorist_uniques

array(['UNKNOWN', 'PEDESTRIAN', 'BICYCLIST', 'MACHINE OPERATOR/RIDER'],
      dtype=object)

In [172]:
non_motorist_substance_abuse = driverdf.non_motorist_substance_abuse
non_motorist_substance_abuse_uniques = non_motorist_substance_abuse.unique()
non_motorist_substance_abuse_uniques

array([nan, 'NONE DETECTED', 'ALCOHOL PRESENT'], dtype=object)

In [173]:
# Fill na's with "UNKNOWN"
driverdf['non_motorist_substance_abuse'] = driverdf['non_motorist_substance_abuse'].fillna("UNKNOWN")
non_motorist_substance_abuse = driverdf.non_motorist_substance_abuse
non_motorist_substance_abuse_uniques = non_motorist_substance_abuse.unique()
non_motorist_substance_abuse_uniques

array(['UNKNOWN', 'NONE DETECTED', 'ALCOHOL PRESENT'], dtype=object)

Now that I see how the data in each column is labeled, I can change the labels for consistency.  (And delete the columns that don't have information I am interested in.

In [174]:
# Delete Columns I don't want to work with---I want to drop 13, but can only drop 8 per command so I will save in a temp

tempdriverdf = driverdf.drop(['person_id', 'vehicle_id', 'driverless_vehicle', ':@computed_region_a9cs_3ed7', ':@computed_region_r648_kzwt', ':@computed_region_vu5j_pcmz', 
              'equipment_problems'], axis=1)

In [175]:
# I will now drop the 5 columns left in the tempdriver dataframe and save it in a new df called driver2df

driver2df = tempdriverdf.drop([':@computed_region_tx5f_5em3',':@computed_region_tx5f_5em3', ':@computed_region_kbsp_ykn9', ':@computed_region_d7bw_bq6x', ':@computed_region_rbt8_3x7n',
                               'off_road_description'], axis = 1)
driver2df



report_number local_case_number               agency_name  \
0      DD5647000F         210007208          Rockville Police   
1      DD5647000F         210007208          Rockville Police   
2      DD5647000C         210007203          Rockville Police   
3     MCP102200FH         210007202  Montgomery County Police   
4     MCP102200FH         210007202  Montgomery County Police   
..            ...               ...                       ...   
995   MCP31740031         210002359  Montgomery County Police   
996   MCP3095002K         210002355  Montgomery County Police   
997   MCP2671002J         210002364  Montgomery County Police   
998   MCP3074004L         210002362  Montgomery County Police   
999   MCP2671002J         210002364  Montgomery County Police   

          acrs_report_type     crash_date_time        route_type  \
0    Property Damage Crash 2021-02-25 10:15:00  Maryland (State)   
1    Property Damage Crash 2021-02-25 10:15:00  Maryland (State)   
2    Property Damage Crash 2021-02-25 09:16:00  Maryland (State)   
3             Injury Crash 2021-02-25 09:16:00  Maryland (State)   
4             Injury Crash 2021-02-25 09:16:00  Maryland (State)   
..                     ...                 ...               ...   
995           Injury Crash 2021-01-19 15:28:00            County   
996  Property Damage Crash 2021-01-19 15:20:00            County   
997  Property Damage Crash 2021-01-19 15:18:00            County   
998  Property Damage Crash 2021-01-19 15:18:00               NaN   
999  Property Damage Crash 2021-01-19 15:18:00            County   

            road_name cross_street_type cross_street_name municipality  ...  \
0          NORBECK RD      Municipality         E GUDE DR    ROCKVILLE  ...   
1          NORBECK RD      Municipality         E GUDE DR    ROCKVILLE  ...   
2    W MONTGOMERY AVE      Municipality         NELSON ST    ROCKVILLE  ...   
3            RIDGE RD            County          BRINK RD          N/A  ...   
4            RIDGE RD            County          BRINK RD          N/A  ...   
..                ...               ...               ...          ...  ...   
995   WATKINS MILL RD            County    APPLE RIDGE RD          N/A  ...   
996         ELDRID DR            County       JOHNSON AVE          N/A  ...   
997         FENTON ST  Maryland (State)  PHILADELPHIA AVE          N/A  ...   
998               NaN           Unknown               NaN          N/A  ...   
999         FENTON ST  Maryland (State)  PHILADELPHIA AVE          N/A  ...   

    speed_limit parked_vehicle vehicle_year vehicle_make      vehicle_model  \
0            50             No         2006       TOYOTA                 4S   
1            50             No         2014       TOYOTA                 4S   
2            30             No         2015          BMW                M4S   
3            40             No         2016       TOYOTA             SIENNA   
4            40             No         2016         FORD              F-150   
..          ...            ...          ...          ...                ...   
995          30             No         2018    CHEVROLET               CRUZ   
996          35             No         2000         FORD  EXPLORER (POSTAL)   
997          35             No         2017      HYUNDAI            ELANTRA   
998          15             No         2001     MERCEDES              4DOOR   
999          35             No         2011         FORD             FUSION   

        latitude     longitude  \
0    39.09043867  -77.12867933   
1    39.09043867  -77.12867933   
2    39.08605617  -77.17062833   
3    39.21003667      -77.2399   
4    39.21003667      -77.2399   
..           ...           ...   
995  39.18419833  -77.21089333   
996     39.06622  -76.99723667   
997  38.98793019   -77.0235645   
998    39.000365  -77.00984667   
999  38.98793019   -77.0235645   

                                           geolocation related_non_motorist  \
0    {'latitude': '3

#### Begin cleaning crash incidents info dataframe

First, begin by examining info about the dataframe.

In [176]:
incidentsdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 51 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   report_number                 1000 non-null   object
 1   local_case_number             1000 non-null   object
 2   agency_name                   1000 non-null   object
 3   acrs_report_type              1000 non-null   object
 4   crash_date_time               1000 non-null   object
 5   hit_run                       1000 non-null   object
 6   route_type                    865 non-null    object
 7   mile_point                    872 non-null    object
 8   mile_point_direction          865 non-null    object
 9   lane_direction                872 non-null    object
 10  lane_number                   1000 non-null   object
 11  number_of_lanes               1000 non-null   object
 12  direction                     872 non-null    object
 13  distance           

In [177]:
# I already know from the driver's info that I'm not interested in columns 40 through 46, so I will drop them now.
incidentsdf.drop(incidentsdf.iloc[:, 40:47], inplace = True, axis = 1)

In [178]:
incidentsdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 44 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   report_number                 1000 non-null   object
 1   local_case_number             1000 non-null   object
 2   agency_name                   1000 non-null   object
 3   acrs_report_type              1000 non-null   object
 4   crash_date_time               1000 non-null   object
 5   hit_run                       1000 non-null   object
 6   route_type                    865 non-null    object
 7   mile_point                    872 non-null    object
 8   mile_point_direction          865 non-null    object
 9   lane_direction                872 non-null    object
 10  lane_number                   1000 non-null   object
 11  number_of_lanes               1000 non-null   object
 12  direction                     872 non-null    object
 13  distance           

#### Convert date-time objects to datetime format

In [179]:
incidentsdf['crash_date_time'] = pd.to_datetime(incidentsdf['crash_date_time'])
incidentsdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 44 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   report_number                 1000 non-null   object        
 1   local_case_number             1000 non-null   object        
 2   agency_name                   1000 non-null   object        
 3   acrs_report_type              1000 non-null   object        
 4   crash_date_time               1000 non-null   datetime64[ns]
 5   hit_run                       1000 non-null   object        
 6   route_type                    865 non-null    object        
 7   mile_point                    872 non-null    object        
 8   mile_point_direction          865 non-null    object        
 9   lane_direction                872 non-null    object        
 10  lane_number                   1000 non-null   object        
 11  number_of_lanes               1

#### Clean up the dataset
Check for duplicates and remove as necessary  
Check for consistency of entries within columns

In [180]:
incidentrprts = incidentsdf.report_number
incidentrprts.nunique()

1000

The number of unique reports is 1000, which means we shouldn't have any duplicates.  I'll double-check with the duplicates command.

In [181]:
duplicateRowsincidentsDF = incidentsdf[incidentsdf.duplicated('report_number')]
duplicateRowsincidentsDF

Empty DataFrame
Columns: [report_number, local_case_number, agency_name, acrs_report_type, crash_date_time, hit_run, route_type, mile_point, mile_point_direction, lane_direction, lane_number, number_of_lanes, direction, distance, distance_unit, road_grade, nontraffic, road_name, cross_street_type, cross_street_name, municipality, at_fault, collision_type, weather, surface_condition, light, traffic_control, driver_substance_abuse, first_harmful_event, second_harmful_event, fixed_oject_struck, junction, intersection_type, intersection_area, road_alignment, road_condition, road_division, latitude, longitude, geolocation, lane_type, off_road_description, related_non_motorist, non_motorist_substance_abuse]
Index: []

[0 rows x 44 columns]

I'll now check the entries in each column, to see if the entries need to be cleaned.

In [182]:
agency_name = incidentsdf.agency_name
agency_uniques = agency_name.unique()
agency_uniques

array(['Rockville Police Departme', 'Montgomery County Police',
       'Maryland-National Capital', 'Gaithersburg Police Depar',
       'Takoma Park Police Depart'], dtype=object)

In [183]:
# Using replace by dictionary to correct the names

incidentsdf['agency_name'] = incidentsdf['agency_name'].replace({'Rockville Police Departme':'Rockville Police', 'Maryland-National Capital':'Maryland-National Capital Police', 
                                                     'Gaithersburg Police Depar':'Gaithersburg Police', 'Takoma Park Police Depart': 'Takoma Park Police'})
agency_uniques = agency_name.unique()
agency_uniques

array(['Rockville Police', 'Montgomery County Police',
       'Maryland-National Capital Police', 'Gaithersburg Police',
       'Takoma Park Police'], dtype=object)

In [184]:
acrs_report_type = incidentsdf.acrs_report_type
acrs_uniques = acrs_report_type.unique()
acrs_uniques

array(['Property Damage Crash', 'Injury Crash', 'Fatal Crash'],
      dtype=object)

In [185]:
hit_run = incidentsdf.hit_run
hit_run_uniques = hit_run.unique()
hit_run_uniques

array(['No', 'Yes'], dtype=object)

In [186]:
route_type = incidentsdf.route_type
route_uniques = route_type.unique()
route_uniques

array(['Maryland (State)', 'County', nan, 'Municipality', 'US (State)',
       'Interstate (State)', 'Other Public Roadway', 'Ramp', 'Government'],
      dtype=object)

In [187]:
# Fill na's with "UNKNOWN"
incidentsdf['route_type'] = incidentsdf['route_type'].fillna("UNKNOWN")
route_type = incidentsdf.route_type
route_uniques = route_type.unique()
route_uniques

array(['Maryland (State)', 'County', 'UNKNOWN', 'Municipality',
       'US (State)', 'Interstate (State)', 'Other Public Roadway', 'Ramp',
       'Government'], dtype=object)

In [188]:
# The numbers won't mean much, but I want to see if there are different ways of representing NA's or unknowns.
mile_point = incidentsdf.mile_point
mile_point_uniques = mile_point.unique()
mile_point_uniques

array(['24.09', '21.29', '0.54', '3.76', '2.28', '12.57', '0', '4.96',
       '0.09', '0.1', '1.14', '3.47', '6.95', '0.04', nan, '3.27',
       '18.81', '1.13', '11.43', '0.53', '6.52', '3.69', '11.29', '0.31',
       '2.23', '4.24', '15.712', '4.32', '0.8', '4.38', '2.03', '20.42',
       '4.21', '0.33', '1.61', '1.06', '0.105', '17.03', '25.33', '14.44',
       '0.6', '2.6', '16.03', '6.36', '0.74', '0.32', '4.11', '0.3',
       '1.4', '0.66', '17.73', '11.84', '9.13', '8.39', '7.87', '4.02',
       '0.36', '7.14', '0.03', '1.1', '8.55', '6.58', '10.85', '11.38',
       '0.52', '1.193', '1.359', '7.75', '5.01', '3.96', '0.61', '10.59',
       '0.955', '1.829', '0.94', '6.21', '10.65', '0.21', '1.45', '4.3',
       '0.85', '0.5', '15.26', '0.27', '0.16', '3.14', '1.8', '0.98',
       '14.74', '11.45', '26.22', '1.83', '8.44', '1.43', '6.174', '0.01',
       '0.76', '2.74', '0.37', '2.65', '0.17', '0.38', '0.86', '1.04',
       '22.72', '0.02', '1.155', '0.65', '2.27', '0.67', '4.64',

In [189]:
incidentsdf['mile_point']=incidentsdf['mile_point'].fillna("UNKNOWN")
mile_point = incidentsdf.mile_point
mile_point_uniques = mile_point.unique()
mile_point_uniques

array(['24.09', '21.29', '0.54', '3.76', '2.28', '12.57', '0', '4.96',
       '0.09', '0.1', '1.14', '3.47', '6.95', '0.04', 'UNKNOWN', '3.27',
       '18.81', '1.13', '11.43', '0.53', '6.52', '3.69', '11.29', '0.31',
       '2.23', '4.24', '15.712', '4.32', '0.8', '4.38', '2.03', '20.42',
       '4.21', '0.33', '1.61', '1.06', '0.105', '17.03', '25.33', '14.44',
       '0.6', '2.6', '16.03', '6.36', '0.74', '0.32', '4.11', '0.3',
       '1.4', '0.66', '17.73', '11.84', '9.13', '8.39', '7.87', '4.02',
       '0.36', '7.14', '0.03', '1.1', '8.55', '6.58', '10.85', '11.38',
       '0.52', '1.193', '1.359', '7.75', '5.01', '3.96', '0.61', '10.59',
       '0.955', '1.829', '0.94', '6.21', '10.65', '0.21', '1.45', '4.3',
       '0.85', '0.5', '15.26', '0.27', '0.16', '3.14', '1.8', '0.98',
       '14.74', '11.45', '26.22', '1.83', '8.44', '1.43', '6.174', '0.01',
       '0.76', '2.74', '0.37', '2.65', '0.17', '0.38', '0.86', '1.04',
       '22.72', '0.02', '1.155', '0.65', '2.27', '0.67', '

In [190]:
mile_point_direction = incidentsdf.mile_point_direction
mpdirection_uniques = mile_point_direction.unique()
mpdirection_uniques

array(['East', 'North', 'South', 'West', nan], dtype=object)

In [191]:
incidentsdf['mile_point_direction'] = incidentsdf['mile_point_direction'].fillna("Unknown")
mile_point_direction = incidentsdf.mile_point_direction
mpdirection_uniques = mile_point_direction.unique()
mpdirection_uniques

array(['East', 'North', 'South', 'West', 'Unknown'], dtype=object)

In [192]:
lane_direction = incidentsdf.lane_direction
lane_direction_uniques = lane_direction.unique()
lane_direction_uniques

array(['West', 'South', 'East', 'North', nan, 'Unknown'], dtype=object)

In [193]:
incidentsdf['lane_direction'] = incidentsdf['lane_direction'].fillna('Unknown')
lane_direction = incidentsdf.lane_direction
lane_direction_uniques = lane_direction.unique()
lane_direction_uniques

array(['West', 'South', 'East', 'North', 'Unknown'], dtype=object)

In [194]:
lane_number = incidentsdf.lane_number
lane_number_uniques = lane_number.unique()
lane_number_uniques

array(['1', '3', '0', '2', '4', '5'], dtype=object)

In [195]:
number_of_lanes = incidentsdf.number_of_lanes
number_of_lanes_uniques = number_of_lanes.unique()
number_of_lanes_uniques

array(['3', '2', '1', '0', '4', '5', '6', '7'], dtype=object)

In [196]:
direction = incidentsdf.direction
direction_uniques = direction.unique()
direction_uniques

array(['East', 'North', 'South', 'West', nan], dtype=object)

In [197]:
incidentsdf['direction'] = incidentsdf['direction'].fillna("Unknown")
direction = incidentsdf.direction
direction_uniques = direction.unique()
direction_uniques

array(['East', 'North', 'South', 'West', 'Unknown'], dtype=object)

In [198]:
distance = incidentsdf.distance
distance_uniques = distance.unique()
distance_uniques

array(['0', '50', '20', '175', '100', nan, '500', '10', '150', '999',
       '15', '0.5', '300', '0.25', '0.4', '200', '30', '1', '25', '362',
       '0.1', '450', '70', '600', '75', '0.2', '185', '900', '350', '250',
       '400', '5', '952', '0.3', '0.6', '60', '0.01', '40', '387', '323',
       '240', '80', '2', '923', '86', '140', '21'], dtype=object)

In [199]:
incidentsdf['distance'] = incidentsdf['distance'].fillna("Unknown")
distance = incidentsdf.distance
distance_uniques = distance.unique()
distance_uniques

array(['0', '50', '20', '175', '100', 'Unknown', '500', '10', '150',
       '999', '15', '0.5', '300', '0.25', '0.4', '200', '30', '1', '25',
       '362', '0.1', '450', '70', '600', '75', '0.2', '185', '900', '350',
       '250', '400', '5', '952', '0.3', '0.6', '60', '0.01', '40', '387',
       '323', '240', '80', '2', '923', '86', '140', '21'], dtype=object)

In [200]:
distance_unit = incidentsdf.distance_unit
distance_unit_uniques = distance_unit.unique()
distance_unit_uniques

array(['FEET', 'MILE', nan, 'UNKNOWN'], dtype=object)

In [201]:
incidentsdf['distance_unit'] = incidentsdf['distance_unit'].fillna("UNKNOWN")
distance_unit = incidentsdf.distance_unit
distance_unit_uniques = distance_unit.unique()
distance_unit_uniques

array(['FEET', 'MILE', 'UNKNOWN'], dtype=object)

In [202]:
road_grade = incidentsdf.road_grade
road_grade_uniques = road_grade.unique()
road_grade_uniques

array(['LEVEL', 'GRADE DOWNHILL', 'HILL UPHILL', 'HILL CREST', nan, 'N/A',
       'ON BRIDGE', 'OTHER', 'DIP SAG', 'UNKNOWN'], dtype=object)

In [203]:
incidentsdf['road_grade'] = incidentsdf['road_grade'].fillna("UNKNOWN")
incidentsdf['road_grade'] = incidentsdf['road_grade'].replace({'N/A': "UNKNOWN"})
road_grade = incidentsdf.road_grade
road_grade_uniques = road_grade.unique()
road_grade_uniques

array(['LEVEL', 'GRADE DOWNHILL', 'HILL UPHILL', 'HILL CREST', 'UNKNOWN',
       'ON BRIDGE', 'OTHER', 'DIP SAG'], dtype=object)

In [204]:
nontraffic = incidentsdf.nontraffic
nontraffic_uniques = nontraffic.unique()
nontraffic_uniques

array(['No', 'Yes'], dtype=object)

In [205]:
cross_street_type = incidentsdf.cross_street_type
cross_street_uniques = cross_street_type.unique()
cross_street_uniques

array(['Municipality', 'County', 'Unknown', 'Maryland (State)', nan,
       'US (State)', 'Ramp', 'Other Public Roadway', 'Service Road',
       'Interstate (State)', 'Government'], dtype=object)

In [206]:
incidentsdf['cross_street_type'] = incidentsdf['cross_street_type'].fillna("Unknown")
cross_street_type = driverdf.cross_street_type
cross_street_uniques = cross_street_type.unique()
cross_street_uniques

array(['Municipality', 'County', 'Unknown', 'Maryland (State)',
       'US (State)', 'Ramp', 'Other Public Roadway', 'Service Road',
       'Interstate (State)'], dtype=object)

In [207]:
municipality = incidentsdf.municipality
municipality_uniques = municipality.unique()
municipality_uniques

array(['ROCKVILLE', 'N/A', nan, 'GAITHERSBURG', 'SOMERSET', 'KENSINGTON',
       'TAKOMA PARK', 'POOLESVILLE', 'FRIENDSHIP HEIGHTS',
       'CHEVY CHASE #4', 'WASHINGTON GROVE', 'CHEVY CHASE #3',
       'LAYTONSVILLE', 'CHEVY CHASE VILLAGE'], dtype=object)

In [208]:
incidentsdf['municipality'] = incidentsdf['municipality'].fillna('N/A')
municipality = incidentsdf.municipality
municipality_uniques = municipality.unique()
municipality_uniques

array(['ROCKVILLE', 'N/A', 'GAITHERSBURG', 'SOMERSET', 'KENSINGTON',
       'TAKOMA PARK', 'POOLESVILLE', 'FRIENDSHIP HEIGHTS',
       'CHEVY CHASE #4', 'WASHINGTON GROVE', 'CHEVY CHASE #3',
       'LAYTONSVILLE', 'CHEVY CHASE VILLAGE'], dtype=object)

In [209]:
at_fault = incidentsdf.at_fault
at_fault_uniques = at_fault.unique()
at_fault_uniques

array(['DRIVER', 'UNKNOWN', 'NONMOTORIST', 'BOTH'], dtype=object)

In [210]:
collision_type = incidentsdf.collision_type
collision_type_uniques = collision_type.unique()
collision_type_uniques

array(['SAME DIRECTION SIDESWIPE', 'SINGLE VEHICLE',
       'STRAIGHT MOVEMENT ANGLE', 'SAME DIR REAR END',
       'HEAD ON LEFT TURN', 'OTHER', 'HEAD ON',
       'SAME DIRECTION RIGHT TURN', 'OPPOSITE DIRECTION SIDESWIPE',
       'UNKNOWN', 'ANGLE MEETS LEFT HEAD ON', 'ANGLE MEETS RIGHT TURN',
       'OPPOSITE DIR BOTH LEFT TURN', 'ANGLE MEETS LEFT TURN',
       'SAME DIR BOTH LEFT TURN', 'SAME DIRECTION LEFT TURN',
       'SAME DIR REND RIGHT TURN', 'SAME DIR REND LEFT TURN', 'N/A'],
      dtype=object)

In [211]:
# replace 'N/A' with UNKNOWN
incidentsdf['collision_type'] = incidentsdf['collision_type'].replace({'N/A': 'UNKNOWN'})
collision_type = incidentsdf.collision_type
collision_type_uniques = collision_type.unique()
collision_type_uniques

array(['SAME DIRECTION SIDESWIPE', 'SINGLE VEHICLE',
       'STRAIGHT MOVEMENT ANGLE', 'SAME DIR REAR END',
       'HEAD ON LEFT TURN', 'OTHER', 'HEAD ON',
       'SAME DIRECTION RIGHT TURN', 'OPPOSITE DIRECTION SIDESWIPE',
       'UNKNOWN', 'ANGLE MEETS LEFT HEAD ON', 'ANGLE MEETS RIGHT TURN',
       'OPPOSITE DIR BOTH LEFT TURN', 'ANGLE MEETS LEFT TURN',
       'SAME DIR BOTH LEFT TURN', 'SAME DIRECTION LEFT TURN',
       'SAME DIR REND RIGHT TURN', 'SAME DIR REND LEFT TURN'],
      dtype=object)

In [212]:
weather = incidentsdf.weather
weather_uniques = weather.unique()
weather_uniques

array(['CLEAR', 'N/A', 'CLOUDY', 'RAINING', 'SLEET', 'SNOW', 'WINTRY MIX',
       'UNKNOWN', 'OTHER', 'BLOWING SNOW', 'FOGGY'], dtype=object)

In [213]:
# replace 'N/A' with 'UNKNOWN'
incidentsdf['weather'] = incidentsdf['weather'].replace({'N/A': 'UNKNOWN'})
weather = incidentsdf.weather
weather_uniques = weather.unique()
weather_uniques

array(['CLEAR', 'UNKNOWN', 'CLOUDY', 'RAINING', 'SLEET', 'SNOW',
       'WINTRY MIX', 'OTHER', 'BLOWING SNOW', 'FOGGY'], dtype=object)

In [214]:
surface_condition = incidentsdf.surface_condition
surface_condition_uniques = surface_condition.unique()
surface_condition_uniques

array(['DRY', 'WET', nan, 'SNOW', 'SLUSH', 'UNKNOWN', 'ICE', 'N/A',
       'WATER(STANDING/MOVING)'], dtype=object)

In [215]:
# replace 'N/A' and fill na's
incidentsdf['surface_condition'] = incidentsdf['surface_condition'].replace({'N/A': 'UNKNOWN'})
incidentsdf['surface_condition'] = incidentsdf['surface_condition'].fillna('UNKNOWN')
surface_condition = incidentsdf.surface_condition
surface_condition_uniques = surface_condition.unique()
surface_condition_uniques

array(['DRY', 'WET', 'UNKNOWN', 'SNOW', 'SLUSH', 'ICE',
       'WATER(STANDING/MOVING)'], dtype=object)

In [216]:
light = incidentsdf.light
light_uniques = light.unique()
light_uniques

array(['DAYLIGHT', 'DARK LIGHTS ON', 'DUSK', 'DARK NO LIGHTS', 'UNKNOWN',
       'DAWN', 'N/A', 'DARK -- UNKNOWN LIGHTING', 'OTHER'], dtype=object)

In [217]:
# Replace 'N/A'
incidentsdf['light'] = incidentsdf['light'].replace({'N/A': 'UNKNOWN'})
light = incidentsdf.light
light_uniques = light.unique()
light_uniques

array(['DAYLIGHT', 'DARK LIGHTS ON', 'DUSK', 'DARK NO LIGHTS', 'UNKNOWN',
       'DAWN', 'DARK -- UNKNOWN LIGHTING', 'OTHER'], dtype=object)